# Mapping the dog epigenome: an inter-tissue and cross-species approach

Keun Hong Son<sup>1,2,3†</sup>, Mark Borris Aldonza<sup>1,2,3†</sup>, A-reum Nam<sup>1,2,3†</sup>, Kang-Hoon Lee<sup>1,3</sup>, Jeong-Woon Lee<sup>1,2,3</sup>, Kyung-Ju Shin<sup>1,3</sup>, Su-Jin Jeong<sup>1,2,3</sup>, Keunsoo Kang<sup>4</sup>, and Je-Yoel Cho<sup>1,2,3*</sup>

<sup>1</sup> Department of Biochemistry, College of Veterinary Medicine, Seoul National University, Seoul, Korea<br>
<sup>2</sup> Comparative Medicine and Disease Research Center (CDRC), Science Research Center (SRC), Seoul National University, Seoul, Korea<br>
<sup>3</sup> BK21 PLUS Program for Creative Veterinary Science Research and Research Institute for Veterinary Science, Seoul National University, Seoul, Korea<br>
<sup>4</sup> Department of Microbiology, College of Natural Sciences, Dankook University, Cheonan, Korea<br>

<sup>†</sup> These authors contributed equally to this work as co-first authors: newhong@snu.ac.kr, borris@snu.ac.kr and arbjlvz@snu.ac.kr<br>
<sup>*</sup> Corresponding author: jeycho@snu.ac.kr

## Introduction

The domestic dog has become a valuable model in exploring multifaceted diseases and biology important for human health. Large-scale dog genome projects produced high-quality draft references but still lack comprehensive annotation of encoded functional elements. Through the integrative next generation sequencing of transcriptomes and chromatin accessibility paired with DNA methylome profiling of 11 adult tissue types, implemented in a cross-species approach, we generated a reference epigenome of a domesticated dog. Using genome orthologues and synthenies, we deciphered the dog’s epigenetic code by defining distinct chromatin states, allowing for genome-wide, integratable data production. We then characterized somatic super-enhancer landscapes and showed that genes mapped on these regions are associated with a broad range of biological and disease traits and are traceable to their tissue-of-origin. Ultimately, we delineated conserved epigenomic changes at the tissue- and species-specific resolutions. Our study provides an epigenomic blueprint of the dog for comparative biology and medical research.

<img src="https://user-images.githubusercontent.com/108702438/177751721-feda7f4b-272d-4750-9208-ac72c187cdb7.jpg" width="600">

## Manuscript and Data availability
Paper (bioRxiv): https://www.biorxiv.org/<br>
Data repository: https://www.ncbi.nlm.nih.gov/geo/<br>
Data browser: [the UCSC_trackhub](http://genome.ucsc.edu/) (Optimizing due to slow speed)<br><br>

## Processed data repository
### RNA contigs
 Transcript contig means the region of the transcript covered by RNA-seq. Therefore, it is possible to study known and novel positions where transcripts are expressed without dependency on genome annotations. (Need to write about data) 

Download: [RNA_Contigs](--)<br>
#### Reference:
- A comparative encyclopedia of DNA elements in the mouse genome, Nature, 2014
- Landscape of transcription in human cells, Nature, 2012<br><br>

### Conserved tissue-specific and species-specific genes
-- (Need to write) <br>
<img src="https://user-images.githubusercontent.com/108702438/177751739-30a13ea6-87ac-4e0f-b260-910bbe31fec3.jpg" width="500">

Download: [Conserved gene lists](--)<br>
#### Reference:
- A comparative encyclopedia of DNA elements in the mouse genome, Nature, 2014
- Comparative transcriptomics in human and mouse, Nature Reviews Genetics, 2017<br><br>

### Chromatin states & Super-enhancer (SE)
 We deciphered the dog’s epigenetic code by defining distinct 13-chromatin state model based on five histone modification marks for 11 tissues, allowing for genome-wide, integratable data production. Then, we characterized somatic super-enhancer landscapes. (Need to write and edit) <br>
<img src="https://user-images.githubusercontent.com/108702438/177751730-c2f1ad14-a5f7-4d4c-a861-764d6dfbb251.jpg" width="250"><br>
<img src="https://user-images.githubusercontent.com/43947916/177108534-54b168a3-280b-49d2-8bc8-a33286eb2c8c.jpg" width="250">

Download: [Chromatin states](https://drive.google.com/drive/folders/1fPqttRt1x6f8RDYpmr8edwPoZyJF_fcv?usp=sharing) / [Super enhancer](https://drive.google.com/drive/folders/1QiNmxW5-nTLRVk8JXh3TNrpKkxzlxOpb?usp=sharing)<br>
#### Reference:
- ChromHMM: automating chromatin-state discovery and characterization, Nature methods, 2012
- Integrative analysis of 111 reference human epigenomes, Nature, 2015<br>
- Master Transcription Factors and Mediator Establish Super-Enhancers at Key Cell Identity Genes, Cell, 2013<br><br>

### SE domain classification and linking with gene
-- (Need to write)  <br>
<img src="https://user-images.githubusercontent.com/108702438/177751763-ebd9a6ff-5990-468b-8ec1-a0c962e5af31.jpg" width="300">

Download: [SE domain-related table](https://github.com/KHworks/dog-reference-epigenome/blob/main/Scripts/Figure%206A_SE-to-gene_linking.py)<br>
#### Reference:
- The chromatin accessibility landscape of primary human cancers, Science, 2018<br><br>
